### PR TITLE
tcp_sink implementation for fluentbit

### PR DESCRIPTION
**Pull request by Proheeler**
Implementation of tcp_sink. Can be used for remote log collectors.

### CODE REVIEW

File: include/spdlog/tcp_sink.h [hunk 1]
```diff
@@ -0,0 +1,47 @@
+
```
Comment (gabime) [inline include/spdlog/tcp_sink.h]: Which platforms does it work on ?

File: include/spdlog/tcp_sink.h [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        }
+
+    }
+protected:
+    void sink_it_(const spdlog::details::log_msg& msg) override
+    {
+        spdlog::memory_buf_t formatted;
+        spdlog::sinks::base_sink<Mutex>::formatter_->format(msg, formatted);
+        std::string message = fmt::to_string(formatted);
+        send(sock , message.c_str() , message.size() , 0 );
```
Comment (gabime) [inline include/spdlog/tcp_sink.h]:
```c++
   send(sock , formatted.data() , formatted.size() , 0 );
```
I more efficient. No need to convert to std::string

File: include/spdlog/tcp_sink.h [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+using namespace std;
+template<typename Mutex>
+class tcp_sink : public spdlog::sinks::base_sink <Mutex>
+{
+public:
+    tcp_sink(std::string address,int port)
+    {
+        if ((sock = socket(AF_INET, SOCK_STREAM, 0)) < 0)
+        {
+            printf("\n Socket creation error \n");
```
Comment (gabime) [inline include/spdlog/tcp_sink.h]: Please throw in case of errors
For example:

```c++
SPDLOG_THROW(spdlog_ex("Socket creation error", errno));
```

Comment (Proheeler) [reply include/spdlog/tcp_sink.h]: Fixed

File: include/spdlog/tcp_sink.h (hunk 2 above)
Comment (Proheeler) [reply include/spdlog/tcp_sink.h]: Fixed)

**Comment by Proheeler** [comment]
This sink works on Linux (tested on Manjaro,Ubuntu,OpenSuse) and MacOS

**Comment by gabime** [comment]
Looks good. Thanks. Please add spdlog license header to to so I cam merge

**Comment by Proheeler** [comment]
> Please also check the return value of send and throw if failed.
Check for send failure done

File: include/spdlog/tcp_sink.h [hunk 4]
```diff
@@ -0,0 +1,50 @@
+// Copyright(c) 2015-present, Gabi Melman & spdlog contributors.
+// Distributed under the MIT License (http://opensource.org/licenses/MIT)
+
+#include <spdlog/spdlog.h>
+#include <spdlog/sinks/base_sink.h>
+#include <sys/socket.h>
+#include <arpa/inet.h>
+#include <unistd.h>
+using namespace std;
```
Comment (gabime) [inline include/spdlog/tcp_sink.h]: Please remove. This wii import all atd symbols to all users

File: include/spdlog/tcp_sink.h [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        spdlog::sinks::base_sink<Mutex>::formatter_->format(msg, formatted);
+        std::string message = fmt::to_string(formatted);
+        send(sock , message.c_str() , message.size() , 0 );
+    }
+
+    void flush_() override
+    {
+    }
+private:
+    int sock;
```
Comment (gabime) [inline include/spdlog/tcp_sink.h]: please add underscore to private member to comply with rest of the spdlog code standard.

Comment (Proheeler) [reply include/spdlog/tcp_sink.h]: Added

**Comment by gabime** [comment]
@Proheeler fyi I merged and committed some fixes.

**Pull request merged**